src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -186,22 +196,108 @@ public void close() throws IOException
       {
         open = false;
         internalFiles.put(name, new Metadata(currOut.getFileNum(), startOffset, currOut.getCurrOffset()));
+        writerCurrentlyInUse = false;
 
         if (bytesWritten != currOut.getCurrOffset() - startOffset) {
           throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
         }
         if (bytesWritten != size) {
           throw new IOException(
               String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
-          );
+              );
+        }
+
+        if(!writerCurrentlyInUse)
+        {
+          mergeWithSmoosher();
+        }
+      }
+    };
+  }
+
+  private void mergeWithSmoosher() throws IOException
+  {
+    //get processed elements from the stack and write.
+    List<File> fileToProcess = new ArrayList<>(files);
+    files = Lists.newArrayList();
+    for(File file: fileToProcess)
+    {
+      add(file);
+      createFiles.remove(file);
+      file.delete();
+    }
+  }
+
+  private SmooshedWriter delegateSmooshedWriter(final String name,final long size) throws IOException
+  {	
+    final File tmpFile = new File(baseDir, name);
+    createFiles.add(tmpFile);
+    return new SmooshedWriter () {
+      private int currOffset = 0;
+      private boolean open = true;
+      private final FileOutputStream out = new FileOutputStream(tmpFile);
+
+      @Override
+      public void close() throws IOException
+      {
+        open = false;
+        out.close();
+        files.add(tmpFile);
+        if(!writerCurrentlyInUse ) {
+          mergeWithSmoosher();
+        }
+      }
+      public int bytesLeft()
+      {
+        return (int) (size - currOffset);
+      }
+
+      @Override
+      public int write(ByteBuffer buffer) throws IOException
+      {
+        FileChannel channel = out.getChannel();
+        int bytesWritten = channel.write(buffer);
+        return  addToOffset(bytesWritten);
+      }
+
+      @Override
+      public int write(InputStream in) throws IOException
+      {
+        return addToOffset(Ints.checkedCast(java.nio.file.Files.copy(in, tmpFile.toPath())));
+      }
+
+      public int addToOffset(int numBytesWritten)
```

Comment:
This method has a really weird contract.  

1. Why is it `public`?
2. Why does it return its incoming argument without any adjustment?

The name of the method makes it seem like this would be returning the total offset instead of the size.  The methods in the non-delegate writer all do this call first with the name "verify" which makes it clear that the intent of the method is to do some state checks...